### PR TITLE
change raw commands into zypper modules

### PR DESCRIPTION
**Pull request by Itxaka**

### CODE REVIEW

**Comment by Itxaka** [comment]
Needs testing to see if zypper dist-upgrade is fixed as per the comments in the code

File: playbooks/openstack-ses_aio_hostconfig.yml [hunk 1]
```diff
@@ -19,12 +19,13 @@
     - name: Ensure all the vars are loaded
       include_vars: "{{ playbook_dir }}/../vars/deploy-on-openstack.yml"
 
-    # TODO(evrardjp): Test if we can include this step in the creation of the node.
-    # In the meantime, we rely on raw ansible module to run task, which causes
-    # the absence of idempotency, and therefore complaining of ansible-lint (301)
     - name: Configure host repositories
-      raw: |
-        zypper ar -f -G {{ deploy_on_openstack_repos_to_configure[item] }} {{ item }}
+      zypper_repository:
```
Comment (aojea) [inline playbooks/openstack-ses_aio_hostconfig.yml]: This has a bug and is not idempotent https://github.com/ansible/ansible/issues/23257 , check if it doesn't affect us

File: playbooks/openstack-osh_hostconfig.yml [hunk 2]
```diff
@@ -5,12 +5,13 @@
     - name: Ensure all the vars are loaded
       include_vars: "{{ playbook_dir }}/../vars/deploy-on-openstack.yml"
 
-    # TODO(evrardjp): Test if we can include this step in the creation of the node.
-    # In the meantime, we rely on raw ansible module to run task, which causes
-    # the absence of idempotency, and therefore complaining of ansible-lint (301)
     - name: Configure host repositories
-      raw: |
-        zypper ar -f -G {{ deploy_on_openstack_repos_to_configure[item] }} {{ item }}
+      zypper_repository:
```
Comment (aojea) [inline playbooks/openstack-osh_hostconfig.yml]: Please check if this is affecting us here https://github.com/ansible/ansible/issues/23257

**Comment by jsuchome** [comment]
Conditional +1 (code is good, needs testing)

**Comment by jsuchome** [comment]
`zypper_repository` seems to work fine when executed again

**Comment by Itxaka** [comment]
tbh reading the module code, its idempotent as it not only checks the repo but also that the options hasnt changed: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/zypper_repository.py#L378

**Comment by Itxaka** [comment]
Issues seems to have been fixed already by https://github.com/ansible/ansible/commit/5464b7156120e9be898e31336bf59b021a64d25f

I also tested this locally with the following play.yml:
```
$ cat play.yml
---
- hosts: localhost
  tasks:
  - name: test adding repo
    zypper_repository:
      name: test repo
      repo: https://download.nvidia.com/opensuse/tumbleweed
      state: present
```

And it didnt fail after the second try:

```
$ ansible-playbook play.yml -v
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] *****************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [localhost]

TASK [test adding repo] **********************************************************************************************************************************************************************
ok: [localhost] => changed=false 
  repodata:
    alias: test repo
    autorefresh: '1'
    enabled: '1'
    gpgcheck: '1'
    name: null
    priority: null
    url: https://download.nvidia.com/opensuse/tumbleweed
  state: present

PLAY RECAP ***********************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0  
```

**Comment by jsuchome** [comment]
What about dist-upgrade?

**Comment by Itxaka** [comment]
> What about dist-upgrade?

Works correctly:

```
$ ansible-playbook play.yml -v
[sudo] password for root: 
Using /etc/ansible/ansible.cfg as config file
/etc/ansible/hosts did not meet host_list requirements, check plugin documentation if this is unexpected
/etc/ansible/hosts did not meet script requirements, check plugin documentation if this is unexpected
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] *****************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [localhost]

TASK [dist-upgrade] **************************************************************************************************************************************************************************
ok: [localhost] => changed=false 
  cmd:
  - /usr/bin/zypper
  - --quiet
  - --non-interactive
  - --xmlout
  - dist-upgrade
  - --auto-agree-with-licenses
  - --no-recommends
  name:
  - '*'
  rc: 0
  state: dist-upgrade
  update_cache: false

PLAY RECAP ***********************************************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=0   
```

**Comment by aojea** [comment]
:thinking: seems that only affects adding repositories with `.repo` files https://github.com/ansible/ansible/issues/23257#issuecomment-291493344

File: playbooks/openstack-osh_hostconfig.yml (hunk 2 above)
Comment (jhesketh) [reply playbooks/openstack-osh_hostconfig.yml]: FYI I see Itxaka's testing here: https://github.com/ansible/ansible/issues/23257#issuecomment-463223472

File: playbooks/openstack-osh_hostconfig.yml [hunk 3]
```diff
@@ -19,19 +20,11 @@
...
-        #zypper ar -f -G https://provo-mirror.opensuse.org/repositories/utilities/SLE_12/ SLE12-utilities;
-        #zypper ar -f -G https://provo-mirror.opensuse.org/repositories/Virtualization:/containers/SLE_12_SP3/ Virtualization_containers;
-        #zypper ar -f -G  http://kirk.provo.novell.com/repositories/SUSE/Products/OpenStack-Cloud/8/x86_64/product/ openstackcloud-8-update
-        zypper refresh && zypper dist-upgrade -y
-      tags:
-        - skip_ansible_lint
+    - name: Zypper dist-upgrade
+      zypper:
+        name: '*'
+        state: dist-upgrade
```
Comment (jhesketh) [inline playbooks/openstack-osh_hostconfig.yml]: Should we `update_cache: yes` here? (equiv of `zypper refresh`)

File: playbooks/openstack-ses_aio_hostconfig.yml [hunk 4]
```diff
@@ -35,19 +36,11 @@
...
-        #zypper ar -f -G https://provo-mirror.opensuse.org/repositories/utilities/SLE_12/ SLE12-utilities;
-        #zypper ar -f -G https://provo-mirror.opensuse.org/repositories/Virtualization:/containers/SLE_12_SP3/ Virtualization_containers;
-        #zypper ar -f -G  http://kirk.provo.novell.com/repositories/SUSE/Products/OpenStack-Cloud/8/x86_64/product/ openstackcloud-8-update
-        zypper refresh && zypper dist-upgrade -y
-      tags:
-        - skip_ansible_lint
+    - name: Zypper dist-upgrade
+      zypper:
+        name: "*"
+        state: dist-upgrade
```
Comment (jhesketh) [inline playbooks/openstack-ses_aio_hostconfig.yml]: ditto

File: playbooks/openstack-osh_hostconfig.yml (hunk 3 above)
Comment (evrardjp) [inline playbooks/openstack-osh_hostconfig.yml]: a zypper dist-upgrade on my machine doesn't actually dist-upgrade.
It runs and do nothing (just run dist-upgrade on the machine later on a machine that has tons of packages to update, and you'll see tons of them).

Is that me, or is it a bug with 12.3?

I moved to re-use zypper refresh && zyper dist-upgrade -y in raw due to that.

**Comment by evrardjp** [comment]
I will run this on the engcloud, see if it works on 12sp3.

**Comment by evrardjp** [comment]
Second run seem to have worked fine...

File: playbooks/openstack-osh_hostconfig.yml (hunk 3 above)
Comment (Itxaka) [reply playbooks/openstack-osh_hostconfig.yml]: This worked correctly on my machine (which is tumbleweed and has ton of packages to update daily)

could it be an ansible version issue? I wuld find that weird as it was introduced in 2017 and it is in versions 2.4 to 2.7 :thinking: 

Let me retest locally

**Comment by evrardjp** [comment]
Just before step 0 here is my package list from dist-upgrade:
```
jevrard-ses:~ # zypper dist-upgrade                                                                             
Warning: You are about to do a distribution upgrade with all enabled repositories. Make sure these repositories a
re compatible before you continue. See 'man zypper' for more information about this command.                    
Loading repository data...                                                                                      
Reading installed packages...                                                                                   
Computing distribution upgrade...                                                                               
                                                                                                                
The following 355 NEW packages are going to be installed:                                                       
  adaptec-firmware adjtimex adwaita-icon-theme atmel-firmware at-spi2-core audit augeas augeas-lenses           
  autoyast2-installation bash-completion bash-doc bind-libs bind-utils boost-license1_54_0 branding-SLE         
  btrfsmaintenance btrfsprogs cantarell-fonts checkmedia coreutils-lang cpio-lang crda cron cronie cups-libs cvs
  cvsps dbus-1-glib dbus-1-python dbus-1-x11 dconf desktop-translations dhcp dhcp-client fontconfig fuse gcr-data
  gcr-lang gcr-prompter gcr-viewer gd gdk-pixbuf-lang gdk-pixbuf-loader-rsvg gdk-pixbuf-query-loaders           
  gio-branding-SLE girepository-1_0 git-cvs git-email git-gui gitk git-svn glib2-lang glib2-tools glib-networking
  glib-networking-lang gpg2-lang grub2-snapper-plugin grub2-systemd-sleep-plugin gsettings-backend-dconf        
  gsettings-desktop-schemas gsettings-desktop-schemas-lang gtk3-data gtk3-lang gtk3-tools gvfs gvfs-backends    
  gvfs-fuse gvfs-lang hicolor-icon-theme initviocons ipw-firmware irqbalance json-glib-lang kernel-firmware     
  keyutils libapr1 libapr-util1 libarchive13 libatasmart4 libatk-1_0-0 libatk-bridge-2_0-0 libatspi0 libauparse0
  libavahi-client3 libavahi-common3 libavahi-glib1 libbluray1 libboost_system1_54_0 libboost_thread1_54_0       
  libbtrfs0 libburn4 libburnia-tools libcairo2 libcairo-gobject2 libcdio14 libcdio_cdda1 libcdio_paranoia1      
  libcolord2 libdatrie1 libdconf1 libdrm2 libdrm_amdgpu1 libdrm_intel1 libdrm_nouveau2 libdrm_radeon1 libelf0   
  libepoxy0 libestr0 libevent-2_0-5 libexif12 libfastjson4 libgbm1 libgc1 libgck-1-0 libgcr-3-1 libgdata22      
  libgdk_pixbuf-2_0-0 libgio-2_0-0 libgirepository-1_0-1 libgmodule-2_0-0 libgnutls28 libgoa-1_0-0              
  libgobject-2_0-0 libgphoto2-6 libgphoto2-6-lang libgraphite2-3 libgtk-3-0 libgudev-1_0-0 libharfbuzz0         
  libhogweed2 libICE6 libisoburn1 libisofs6 libjasper1 libjbig2 libjpeg8 libjson-glib-1_0-0 liblcms2-2          
  libldapcpp1 libLLVM liblockdev1 liblogging0 libltdl7 libmozjs-17_0 libmtp9 libnettle4 libnuma1 liboauth0      
  libpango-1_0-0 libpci3 libpciaccess0 libpixman-1-0 libpolkit0 librest0 librsvg-2-2 libsecret-1-0 libsecret-lang
  libserf-1-1 libSM6 libsnapper4 libsoup-2_4-1 libsoup-lang libstorage7 libstorage-ruby libthai0 libthai-data 
  libtidyp-1_04-0 libtiff5 libtirpc3 libtirpc-netconfig libudisks2-0 libvpx1 libX11-6 libX11-data libX11-xcb1   
  libXau6 libxcb1 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-render0 libxcb-shm0 libxcb-sync1  libxcb-xfixes0 libXcomposite1 libXcursor1 libXdamage1 libXext6 libXfixes3 libXft2 libXi6 libXinerama1         
  libxml2-tools libXmuu1 libXpm4 libXrandr2 libXrender1 libxshmfence1 libXss1 libXt6 libXtst6 libXxf86vm1 lockdev  lsof lsscsi lvm2 Mesa Mesa-libEGL1 Mesa-libGL1 Mesa-libglapi0 mpt-firmware nfs-client nfsidmap openssh-askpass
  openssh-helpers os-prober pciutils pciutils-ids perl-Authen-SASL perl-Config-Crontab perl-DBD-SQLite perl-DBI 
  perl-Encode-Locale perl-File-Listing perl-gettext perl-HTML-Parser perl-HTML-Tagset perl-HTML-Tidy            
  perl-HTTP-Cookies perl-HTTP-Daemon perl-HTTP-Date perl-HTTP-Message perl-HTTP-Negotiate perl-IO-HTML          
  perl-IO-Socket-SSL perl-libwww-perl perl-LWP-MediaTypes perl-LWP-Protocol-https perl-Net-Daemon perl-Net-DBus 
  perl-Net-HTTP perl-Net-SMTP-SSL perl-Net-SSLeay perl-Parse-RecDescent perl-PlRPC perl-Term-ReadKey            
  perl-Tie-IxHash perl-TimeDate perl-URI perl-WWW-RobotRules perl-X11-Protocol perl-X500-DN                     
  perl-XML-NamespaceSupport perl-XML-Parser perl-XML-SAX perl-XML-SAX-Base perl-XML-SAX-Expat perl-XML-Simple   
  perl-XML-Twig perl-XML-XPath perl-XML-XPathEngine polkit polkit-default-privs postfix python-gobject          
  python-gobject-cairo python-gobject-Gdk python-pam python-pycurl python-PySocks python-strict-tls-check       
  python-Twisted python-zope.interface readline-doc rollback-helper rpcbind rsyslog ruby                        
  ruby2.1-rubygem-abstract_method ruby2.1-rubygem-cfa ruby2.1-rubygem-cfa_grub2 ruby2.1-rubygem-cheetah         
  ruby2.1-rubygem-fast_gettext ruby2.1-rubygem-ruby-augeas ruby2.1-rubygem-ruby-dbus sed-lang shared-mime-info  
  snapper snapper-zypp-plugin subversion subversion-perl syslinux syslog-service systemd-bash-completion tar-lang
  tcl tcsh time tk traceroute typelib-1_0-Atk-1_0 typelib-1_0-GdkPixbuf-2_0 typelib-1_0-Gtk-3_0                 
  typelib-1_0-Pango-1_0 ucode-intel udisks2 udisks2-lang unzip util-linux-lang vim-data w3m                     
  wallpaper-branding-SLE which wireless-regdb xauth xdg-utils xhost xz-lang yast2 yast2-add-on yast2-bootloader 
  yast2-branding-SLE yast2-core yast2-country yast2-country-data yast2-firewall yast2-hardware-detection        
  yast2-installation yast2-ldap yast2-migration yast2-network yast2-online-update yast2-packager yast2-pam      
  yast2-perl-bindings yast2-pkg-bindings yast2-proxy yast2-registration yast2-ruby-bindings yast2-security      
  yast2-services-manager yast2-slp yast2-snapper yast2-storage yast2-theme-SLE yast2-transfer yast2-trans-stats 
  yast2-update yast2-users yast2-xml yast2-ycp-ui-bindings zypper-lifecycle-plugin zypper-log                   
  zypper-migration-plugin

The following 12 packages have no support information from their vendor:                                        
  cvsps git-cvs git-email git-gui gitk git-svn libserf-1-1 libtidyp-1_04-0 perl-HTML-Tidy perl-Net-SMTP-SSL     
  subversion subversion-perl

355 new packages to install.
Overall download size: 148.4 MiB. Already cached: 0 B. After the operation, additional 571.8 MiB will be used.
```

I seriously doubt the dist-upgrade step worked.

**Comment by Itxaka** [comment]
Locally it works as expected:

```
PLAY [localhost] *****************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************
ok: [localhost]

TASK [zypper dist upgrade] *******************************************************************************************************************************************************************
changed: [localhost] => changed=true 
  cmd:
  - /usr/bin/zypper
  - --quiet
  - --non-interactive
  - --xmlout
  - dist-upgrade
  - --auto-agree-with-licenses
  - --no-recommends
  name:
  - '*'
  rc: 0
  state: dist-upgrade
  stderr: ''
  stderr_lines: []
  stdout: |-
    <?xml version='1.0'?>
    <stream>
    <install-summary download-size="83091310" space-usage-diff="4109" packages-to-change="2">
    <to-upgrade>
    <solvable type="package" name="google-chrome-stable" edition="72.0.3626.109-1" arch="x86_64" edition-old="72.0.3626.96-1" arch-old="x86_64" summary="Google Chrome">
    <description>The web browser from Google
  
    Google Chrome is a browser that combines a minimal design with sophisticated technology to make the web faster, safer, and easier.</description></solvable>
    <solvable type="package" name="wine-32bit" edition="4.1-599.3" arch="x86_64" edition-old="4.1-599.1" arch-old="x86_64" summary="An MS Windows Emulator">
    <description>Wine is a compatibility layer capable of running Windows
    applications. Instead of simulating internal Windows logic like a
    virtual machine or emulator, Wine translates Windows API calls into
    POSIX calls on-the-fly, eliminating the performance and memory
    penalties of other methods and allowing you to cleanly integrate
    Windows applications into your desktop.
  
  
  
  
  
  
  
  
  
  
    You can run your Windows executables with it and write your Windows
    programs under Linux and link against the WINE libraries. It is not
    necessary to have a Windows installation to run WINE.
  
    Refer to /usr/share/doc/packages/wine/README.SUSE. There is more
    documentation available in that directory. Read &apos;man wine&apos; for further
    information.
  
    You can invoke wine by entering &apos;wine program.exe&apos;. Configure it by
    running &apos;winecfg&apos;.</description></solvable>
    </to-upgrade>
    </install-summary>
    <prompt id="0">
    <text>Continue?</text>
    <option default="1" value="y" desc="Yes, accept the summary and proceed with installation/removal of packages."/>
    <option value="n" desc="No, cancel the operation."/>
    <option value="v" desc="Toggle display of package versions."/>
    <option value="a" desc="Toggle display of package architectures."/>
    <option value="r" desc="Toggle display of repositories from which the packages will be installed."/>
    <option value="m" desc="Toggle display of package vendor names."/>
    <option value="d" desc="Toggle between showing all details and as few details as possible."/>
    <option value="g" desc="View the summary in pager."/>
    </prompt>
    <download url="http://dl.google.com/linux/chrome/rpm/stable/x86_64/google-chrome-stable-72.0.3626.109-1.x86_64.rpm" percent="-1" rate="-1"/>
    <download url="http://dl.google.com/linux/chrome/rpm/stable/x86_64/google-chrome-stable-72.0.3626.109-1.x86_64.rpm" percent="0" rate="2514"/>
    <download url="http://dl.google.com/linux/chrome/rpm/stable/x86_64/google-chrome-stable-72.0.3626.109-1.x86_64.rpm" percent="51" rate="29586602"/>
    <download url="http://dl.google.com/linux/chrome/rpm/stable/x86_64/google-chrome-stable-72.0.3626.109-1.x86_64.rpm" rate="29586602" done="1"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="-1" rate="-1"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="0" rate="2276"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="21" rate="2735075"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="45" rate="3555552"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="63" rate="4268844"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="77" rate="4212707"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="87" rate="4114195"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" percent="96" rate="4139167"/>
    <download url="http://download.opensuse.org/repositories/Emulators/openSUSE_Tumbleweed/x86_64/wine-32bit-4.1-599.3.x86_64.rpm" rate="4067627" done="1"/>
    </stream>
  stdout_lines: <omitted>
  update_cache: false

PLAY RECAP ***********************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0 
```

Not sure why it would fail on SES. Let me retest this and confirm

File: playbooks/openstack-osh_hostconfig.yml (hunk 3 above)
Comment (Itxaka) [reply playbooks/openstack-osh_hostconfig.yml]: Good point, let me check the modules code to see if it is compatible

Comment (evrardjp) [reply playbooks/openstack-osh_hostconfig.yml]: Could we try to remove that whole task, see if things fail anyway? I have the impression salt is doing the dist-upgrade anyway.

**Comment by Itxaka** [comment]
oh so it upgrades but it does not install new packages:

before dist-upgrade runs:

```
179 packages to upgrade, 355 new, 1 to remove, 1 to change arch.
```

So it upgrades the packages but does not install the new ones. We may need an extra flag for that

**Comment by Itxaka** [comment]
@evrardjp So this seems to be working as expected, i.e. update the system VS upgrade the system with `--no-recommends` enabled by default

If you want and we need it, we can probably set the `disable_recommends` option, but that would install *new* packages which I dont think its good for a clean image (even if seadep does it afterwards, they may  do it for their own reasons)

Thoughs?

**Comment by evrardjp** [comment]
I think we can keep your patch code, and see reliability over time.

**Comment by evrardjp** [comment]
I will try this on top of my leap15 patch, to see how it goes.

File: playbooks/openstack-osh_hostconfig.yml [hunk 5]
```diff
@@ -19,19 +20,12 @@
...
+        update_cache: yes
```
Comment (evrardjp) [inline playbooks/openstack-osh_hostconfig.yml]: could you add `extra_args_precommand: "--gpg-auto-import-keys"` ? I am testing this right now, as the key is not trusted after import on the previous step.

Comment (evrardjp) [reply playbooks/openstack-osh_hostconfig.yml]: Worked with extra_args_precommand.

Comment (Itxaka) [reply playbooks/openstack-osh_hostconfig.yml]: done!

**Comment by Itxaka** [comment]
\o/

**Pull request merged**